hie-build-plugin/Haskell/Ide/BuildPlugin.hs

```diff
@@ -216,16 +302,31 @@ flagToJSON f = object ["name" .= ((\(FlagName s) -> s) $ flagName f), "descripti
 
 -----------------------------------------------
 
-buildDirectory :: CommandFunc ()
-buildDirectory = CmdSync $ \ctx req -> withCommonArgs ctx req $ do
+data BuildParams = BP {
+         -- common params. horrible
```

Comment:
Because of the way the plugin stuff works you need to specify a type that comes from the client, and is decoded from JSON.

The reason there are three types is that the function passed in to `withCommonArgs` expects the common arguments in a structure that is always populated, so cannot have `Maybe` values. But some common params are optional, so the `CommonParams` structure exists to manage this.

The `BuildParams` are an extension of the `CommonParams`, but we need to give a single type that can be serialised through JSON in a natural way.  Hence the duplication.

Thinking about it, we might be able to make completely off-the-wall JSON instances where `BuildParams` has a `CommonParams` field and then the extension, but it maps out to a flat JSON structure as expected on the wire.

@wz1000 thoughts?